src/storage/src/upsert.rs

```diff
@@ -670,259 +850,341 @@ where
-                match event {
-                    AsyncEvent::Data(_cap, data) => {
-                        events.extend(data.into_iter().filter_map(|((key, value), ts, diff)| {
-                            if !resume_upper.less_equal(&ts) {
-                                Some((key, value, diff))
-                            } else {
-                                None
-                            }
-                        }))
-                    }
-                    AsyncEvent::Progress(upper) => {
-                        snapshot_upper = upper;
-                    }
-                };
-            }
 
-            for (_, value, diff) in events.iter_mut() {
-                if let Err(UpsertError::Value(ref mut err)) = value {
-                    // If we receive a legacy error in the snapshot we will keep a note of it but
-                    // insert a non-legacy error in our state. This is so that if this error is
-                    // ever retracted we will correctly retract the non-legacy version because by
-                    // that time we will have emitted the error correction, which happens before
-                    // processing any of the new source input.
-                    if err.is_legacy_dont_touch_it {
-                        legacy_errors_to_correct.push((err.clone(), diff.clone()));
-                        err.is_legacy_dont_touch_it = false;
-                    }
-                }
-            }
+        loop {
+            tokio::select! {
+                Some(persist_event) = persist_input.next() => {
 
-            match state
-                .consolidate_snapshot_chunk(
-                    events.drain(..),
-                    PartialOrder::less_equal(&resume_upper, &snapshot_upper),
-                )
-                .await
-            {
-                Ok(_) => {
-                    if let Some(ts) = snapshot_upper.clone().into_option() {
-                        // As we shutdown, we could ostensibly get data from later than the
-                        // `resume_upper`, which we ignore above. We don't want our output capability to make
-                        // it further than the `resume_upper`.
-                        if !resume_upper.less_equal(&ts) {
-                            snapshot_cap.downgrade(&ts);
-                            output_cap.downgrade(&ts);
+                    // Buffer as many events as possible. This should be
+                    // bounded, as new data can't be produced in this worker
+                    // until we yield to timely.
+                    let persist_events = [persist_event]
+                        .into_iter()
+                        .chain(std::iter::from_fn(|| persist_input.next().now_or_never().flatten()));
+
+                    // Read away as much input as we can.
+                    for persist_event in persist_events {
+                        tracing::trace!(?persist_event, "persist input");
```

Comment:
I left it in on purpose, because it was very useful for me when debugging, and we normally don't even log at DEBUG level in prod. But I can take these out if you think it's to risky to leave them in at TRACE